### ISSUE TITLE
machine_info.py: warnings do not show up during tests

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#668]
https://github.com/conbench/conbench/blob/b524d43468bc4cfce04c7aba59bc2392a595355f/conbench/machine_info.py#L100

This indirection is not of clear value:
```
class GitParseWarning(RuntimeWarning):
    pass
```

This indirection makes it so that these warnings do not show during a pytest session (and possibly never by default, I did not test this).

`RuntimeWarning` does not seem to be the right base to derive from when one wants to emit warnings from applications.

Leaving this as the _default_ https://docs.python.org/3/library/exceptions.html#UserWarning is more useful. This shows up during a pytest session:

```
conbench/tests/benchmark/test_runner.py::test_runner_simple_benchmark
  /app/conbench/machine_info.py:102: UserWarning: error in github_info(): probably not in a git repo; returning {}
    warnings.warn(
```

**Comment by jgehrcke** [comment]
I just saw that these warnings were explicitly silenced during a pytest session:

https://github.com/conbench/conbench/blob/b524d43468bc4cfce04c7aba59bc2392a595355f/conbench/tests/benchmark/test_runner.py#L16

This is like a trap for fellow developers:
- have a test require a feature to work
- make that feature emit a specific warning when it does not work, but otherwise not throw an exception
- then silence that specific warning during tests

I am sure it wasn't meant that way, but this feels like an easteregg :egg:  :D :male_detective:

**Comment by austin3dickey** [comment]
Oh I did this! Here were my thoughts:

1. I want to suppress expected warnings in the pytest summary, to reduce noise
2. I want to see unexpected warnings in the pytest summary, so developers can act on them
3. `github_info()` tries to read certain git information from the local git setup, and emits a warning if it fails to do so for some reason. One instance that it will fail to read the correct information is if it's running on a detached HEAD.
4. Sometimes we run tests on a local computer on a tracked branch, where `github_info()` would succeed. In GitHub Actions, it always runs tests on a detached HEAD, where `github_info()` would fail and warn.
5. Because of (3) and (4), sometimes we'll see these warnings in the test suite and sometimes we won't. This fails point (1). So we have to catch the warnings in the test suite somehow. If we don't catch them (like you tried in #667) we will see expected noise like [this](https://github.com/conbench/conbench/actions/runs/4084987390/jobs/7046718157#step:3:2471) in the pytest summary.
7. We can't catch them with [pytest.warns()](https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html#warns) because of (4). Sometimes the warning won't appear, and that's OK.
8. We can't suppress all warnings in pytest because of (2).
9. Therefore we have to filter out these specific warnings in the test suite using [pytest.mark.filterwarnings](https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html#pytest-mark-filterwarnings). (I found the best way to filter out only these warnings — and no other warnings — is to create a custom warning class and use that for the filter.)

If you have other ideas, let me know. I suppose we could try to mock out a local git repository that's the same no matter if you run tests on a laptop or GitHub Actions. That seemed pretty hard at the time, but maybe it wouldn't be as hard with the Dockerized setup.

Note that, currently, we never test that `github_info()` returns a certain expected value, because of this variability in the environment. We just let it be called as a side effect during some tests, and never check the output. Another possibility is to ensure `github_info()` never is called during the test suite, but I didn't like that as much.

**Comment by jgehrcke** [comment]
Thanks for the thoughtful reply, and for describing the line of thinking. I see that you tried to bring a lot of structure into this, and I appreciate this.

We need to align on the following:

> we never test that github_info() returns a certain expected value [...] We just let it be called [...] and never check the output.

Mhm! This is not true as far as I understood code. The test failures described in https://github.com/conbench/conbench/pull/667#issuecomment-1415879998 were because the output of `github_info()` was not as the tests expected. The unexpected output did not pass schema validation. Schema validation is expected to return an empty dict, but it returned  `{'github': {'commit': ['Missing data for required field.'], 'repository': ['Missing data for required field.']}}` instead. Because `github_info()` returned `{}`. (please let me know if this analysis is not correct)

This needs a careful look. The error at the surface was:
```
        benchmark = SimpleBenchmark()
        [(result, output)] = benchmark.run(run_id=None, batch_id=None)
>       assert not BenchmarkFacadeSchema.create.validate(result)
E       AssertionError: assert not {'github': {'commit': ['Missing data for required field.'], 'repository': ['Missing data for required field.']}}
```

It took me ~20 minutes to translate this error into "ah, the code expects a git repo in the current working dir, and I removed this dependency unknowingly".

One of the suppressed warnings would have helped reduce that time. Yes. But the problem that matters more is the implicit dependency, not so much the suppressed warning.

So. We have two specific issues at hand I think:

1. "warning did not show up in logs, unexpectedly" (cf. your point (2))
2. "test suite has implicit dependency on git repo in current working directory, fails with misleading errors when this dependency is not met"

For developer productivity it is more important to address (2) than (1). Because:

- when the test suite fails hard, we want an error message that saves time, quickly reveals what is wrong.
- warnings are a subjective topic and provide a blurry interface where it's sometimes not even clear who the interface consumer is. also, the rigor of this interface is usually not tested, i.e. not actually present.

I should have made an issue about (2) instead of (1).

So. Zooming out for a moment, I think this is a productive thought for the actual challenge at hand: the test suite expects that the current working directory contains a git repository (maybe a specific one). If that is the case then it should explicitly check for this dependency. If it is not met it should emit a corresponding error message. This can be as simple as `assert os.path.exists('.git'), ".git directory in cwd expected"`, would have saved me 20 minutes. As far as I have understood, this would codify an assumption (which is always a good idea to help fellow debuggers).

I hope up to here we can pretty much agree.

Given that you said "we never test that `github_info()` returns a certain expected value" I trust that this dependency / assumption was somehow added unknowingly.

Now, the slightly more subjective topic of warnings:

> I want to suppress expected warnings

Distinguishing expected warnings from unexpected warnings and suppressing unexpected warnings for reducing noise sounds like an honorable goal. But it also is somehow against various engineering intuitions, it smells like an energy investment for little or even hurting return. Yes _actual_ invariants (enforced by code) help. Being able to make precise statements about state. With weakly derived assumptions (not enforced by code) the distinction between "expected" and "unexpected" becomes error-prone. The scenario that I ran into (my point (1), your point (2)) is I think a manifestation of this.

I want to go back to your

> We can't suppress all warnings in pytest because of (2).
> I want to see unexpected warnings in the pytest summary, so developers can act on them

Yes, that makes sense! But this makes me think that instead of "seeing unexpected warnings in pytest" I want to see an error in pytest :-). I think that summarizes things. If the test suite is green and shows many warnings, it's still green.

So, expected vs. unexpected warnings are a source of confusion. Adding rigor on top of that means turning warnings into actual errors.

Some consider exchanges like this to be a waste of time, but I am actually grateful that we have this. Hope you can agree :). I feel like I have learned a lot here.

For #667 I think I will revert all changes related to warnings. And just add something like `assert os.path.exists('.git'), ".git directory in cwd expected"` so that the next person who touches this has an easier time understanding the problem.

> Another possibility is to ensure `github_info()` never is called during the test suite, but I didn't like that as much.

Sounds like the interface guarantees for this function need some work :). Once we specify / document the interface with precision, we can and should test it somewhere, properly.

**Comment by austin3dickey** [comment]
>Given that you said "we never test that github_info() returns a certain expected value" I trust that this dependency / assumption was somehow added unknowingly.

Yes, absolutely! I didn't realize the test suite depended on having a local git repo available; that's unintentional. The intention was that `github_info()` is untested, and tests shouldn't fail because of it. That seems to be untrue. The ideal state is that we do test it in a few controlled environments, and other tests don't fail outside of those environments.

>Adding rigor on top of that means turning warnings into actual errors.

Yep I've done this in past projects. I like the extra rigor that "warnings fail CI" creates. But we've never journeyed that far in this repo.
